plansys2_executor/src/plansys2_executor/ExecutorClient.cpp

```diff
@@ -23,67 +23,146 @@
+    rclcpp::spin_some(node_);
 
-  if (!this->execute_plan_client_ptr_) {
-    RCLCPP_ERROR(node_->get_logger(), "Action client not initialized");
+    if (!goal_result_available_) {
+      return true;  // Plan not finished
+    }
   }
 
-  if (!this->execute_plan_client_ptr_->wait_for_action_server(std::chrono::seconds(10))) {
-    RCLCPP_ERROR(node_->get_logger(), "Action server not available after waiting");
-    return false;
+  switch (result_.code) {
+    case rclcpp_action::ResultCode::SUCCEEDED:
+      RCLCPP_INFO(node_->get_logger(), "Plan Succeded");
+      break;
+
+    case rclcpp_action::ResultCode::ABORTED:
+      RCLCPP_WARN(node_->get_logger(), "Plan Aborted");
+      break;
+
+    case rclcpp_action::ResultCode::CANCELED:
+      RCLCPP_INFO(node_->get_logger(), "Plan Cancelled");
+      break;
+
+    default:
+      throw std::logic_error("ExecutorClient::executePlan: invalid status value");
   }
 
-  auto goal_msg = ExecutePlan::Goal();
+  executing_plan_ = false;
+  goal_result_available_ = false;
+
+  return false;  // Plan finished
+}
+
 
+bool
+ExecutorClient::on_new_goal_received()
+{
   auto send_goal_options = rclcpp_action::Client<ExecutePlan>::SendGoalOptions();
 
   send_goal_options.feedback_callback =
     std::bind(&ExecutorClient::feedback_callback, this, _1, _2);
 
   send_goal_options.result_callback =
     std::bind(&ExecutorClient::result_callback, this, _1);
-  auto goal_handle_future = this->execute_plan_client_ptr_->async_send_goal(
-    goal_msg, send_goal_options);
 
-  if (rclcpp::spin_until_future_complete(node_, goal_handle_future) !=
-    rclcpp::executor::FutureReturnCode::SUCCESS)
+  auto future_goal_handle = action_client_->async_send_goal(goal_, send_goal_options);
+
+  if (rclcpp::spin_until_future_complete(
+      node_->get_node_base_interface(), future_goal_handle, 3s) !=
+    rclcpp::FutureReturnCode::SUCCESS)
   {
     RCLCPP_ERROR(node_->get_logger(), "send_goal failed");
     return false;
   }
 
-  auto goal_handle = goal_handle_future.get();
-  if (!goal_handle) {
-    RCLCPP_ERROR(
-      node_->get_logger(), "ExecutorClient: Plan execution was rejected by the action server");
+  goal_handle_ = future_goal_handle.get();
+  if (!goal_handle_) {
+    RCLCPP_ERROR(node_->get_logger(), "Goal was rejected by the action server");
     return false;
   }
 
   return true;
 }
 
+bool
+ExecutorClient::should_cancel_goal()
+{
+  if (!executing_plan_) {
+    return false;
+  }
+
+  rclcpp::spin_some(node_);
+  auto status = goal_handle_->get_status();
+
+  return status == action_msgs::msg::GoalStatus::STATUS_ACCEPTED ||
+         status == action_msgs::msg::GoalStatus::STATUS_EXECUTING;
+}
+
+void
+ExecutorClient::cancel_plan_execution()
+{
+  if (should_cancel_goal()) {
+    auto future_cancel = action_client_->async_cancel_goal(goal_handle_);
+    if (rclcpp::spin_until_future_complete(
+        node_->get_node_base_interface(), future_cancel, 3s) !=
```

Comment:
Why are those 3 second for ? Is this possible to have a plan cancellation that last up to 3 seconds ?

Reply:
By the way, it is a fantastic work.

Reply:
Well, it's an arbitrary time. It will probably take less, but I have no clear criteria for timeouts until tested in robots with poor connection...

Reply:
In this case isn't it imaginable to make it a parameter with a default value ?

Reply:
Yes, we should make it a parameter in any moment